src/main/java/edu/harvard/iq/dataverse/search/SearchServiceBean.java

```diff
@@ -1006,6 +1007,14 @@ private String getPermissionFilterQuery(DataverseRequest dataverseRequest, SolrQ
         // Yes, see if GuestUser is part of any groups such as IP Groups.
         // ----------------------------------------------------
         if (user instanceof GuestUser) {
+            if (FeatureFlags.AVOID_EXPENSIVE_SOLR_JOIN.enabled()) {
```

Comment:
Could you just check to see if groupsFromProviders (or any other reason to not use the public_object) here?

Reply:
@qqmyers I'm not sure I follow.

My goal is to do a quick check to see if a feature flag is set or not. If set, return early without a join.

You're proposing something else?

Reply:
If the flag and no groupsFromProviders  (or the subset of those that are IP groups) then skip the join. Guessing that would make this work in all circumstances?

Reply:
Are we only talking about making an extra effort NOT to treat members of IP groups as plain guest users here? (I remember very little about how IP groups are supposed to work; I'll let you guys figure it out)

Reply:
@qqmyers what is the problem you are trying to solve? What are you concerned about?

Reply:
"However, dvObjects that would have been found by IP Groups will no longer be found." If a user is a member of an IP group, why can't you skip adding the `SearchFields.PUBLIC_OBJECT + ":" + true;` test and just do the normal join? That doesn't give the speedup if everyone is in an IP group, but if that's only a subset of users, you'd get the speedup for others and not break IP groups.

Reply:
Ok, I think I get it. I haven't tested the code below, but is it closer to what you're thinking?

```java
// ----------------------------------------------------
// (1) Is this a GuestUser?
// Yes, see if GuestUser is part of any groups such as IP Groups.
// ----------------------------------------------------
if (user instanceof GuestUser) {
    String groupsFromProviders = "";
    Set<Group> groups = groupService.collectAncestors(groupService.groupsFor(dataverseRequest));
    StringBuilder sb = new StringBuilder();
    for (Group group : groups) {
        logger.fine("found group " + group.getIdentifier() + " with alias " + group.getAlias());
        String groupAlias = group.getAlias();
        if (groupAlias != null && !groupAlias.isEmpty()) {
            sb.append(" OR ");
            // i.e. group_builtIn/all-users, ip/ipGroup3
            sb.append(IndexServiceBean.getGroupPrefix()).append(groupAlias);
        }
    }
    groupsFromProviders = sb.toString();
    logger.fine("groupsFromProviders:" + groupsFromProviders);
    String guestWithGroups = "{!join from=" + SearchFields.DEFINITION_POINT + " to=id}" + SearchFields.DISCOVERABLE_BY + ":(" + IndexServiceBean.getPublicGroupString() + groupsFromProviders + ")";
    logger.fine(guestWithGroups);
    // Check if any groups from providers were found.
    if ("".equals(groupsFromProviders)) {
        if (FeatureFlags.AVOID_EXPENSIVE_SOLR_JOIN.enabled()) {
            /**
             * Now that we know that there are no groupsFromProviders
             * (such as IP Groups) involved, instead of doing an
             * expensive join, narrow down to only public objects. This
             * field is indexed on the content document itself, rather
             * than a permission document.
             */
            return SearchFields.PUBLIC_OBJECT + ":" + true;
        }
    }
    /**
     * Return the expensive join if groups from providers (IP Groups)
     * were found or if the feature flag isn't enabled.
     */
    return guestWithGroups;
}
```